### ISSUE TITLE
Too many open files during terminate and replace node (scale test)

### ISSUE THREAD

**Issue by soyacz** [scylladb/scylla#10462]
## Installation details

Kernel Version: 5.13.0-1021-aws
Scylla version (or git commit hash): `5.0~rc3-20220406.f92622e0d` with build-id `2b79c4744216b294fdbd2f277940044c899156ea`
Cluster size: 4 nodes (i3.metal)

Scylla Nodes used in this run:
- longevity-many-clients-4h-5-0-db-node-4133baf8-5 (54.216.121.122 | 10.0.1.85) (shards: 70)
- longevity-many-clients-4h-5-0-db-node-4133baf8-4 (54.246.175.60 | 10.0.2.106) (shards: 70)
- longevity-many-clients-4h-5-0-db-node-4133baf8-3 (52.50.93.24 | 10.0.0.87) (shards: 70)
- longevity-many-clients-4h-5-0-db-node-4133baf8-2 (34.254.180.120 | 10.0.2.251) (shards: 70)
- longevity-many-clients-4h-5-0-db-node-4133baf8-1 (3.251.71.31 | 10.0.3.173) (shards: 70)

OS / Image: `ami-07835983d717b1ea3` (aws: eu-west-1)

Test: `scale-many-clients-4h-test`
Test id: `4133baf8-c193-41bc-a1c6-1f18c434701d`
Test name: `scylla-5.0/scale/scale-many-clients-4h-test`
Test config file(s):

- [longevity-many-clients-4h.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/1cdf92e8270cc05282286e2df33d21e59bc3b712/test-cases/scale/longevity-many-clients-4h.yaml)

## Issue description

When running a scale test ( 4xi3.metal db nodes, 50 loaders), during terminate_and_replace_node nemesis, in the added node logs we can see errors like (RF=3):
```
[shard 0] storage_service - replace[83ecc76b-2dd3-40b4-9d75-3a58480cb689]: Abort replace operation started, replace_nodes={{ 10.0.0.87 -> 10.0.1.85}}, sync_nodes=[10.0.1.85, 10.0.2.106, 10.0.2.251, 10.0.3.173], ignore_nodes=[]: std::runtime_error ({shard 0: std::filesystem::__cxx11::filesystem_error (error system:24, filesystem error: open failed: Too many open files [/var/lib/scylla/data/keyspace40/standard1-b1405a20c79211ecb8964131e1a93ba3/md-21280-big-Index.db])
```
when looking into monitoring, we can see rapid growth of files no on that node:
![open_files_issue](https://user-images.githubusercontent.com/18242288/165970257-0f9c5a6d-2f81-4e9d-a7e9-3ed48949673b.png)

During that time, write stress on all 50 loaders crashed - dropping write load to 0 (over ~15mins). Example error from loader:
```
WARN  10:59:10,882 Error creating netty channel to 10.0.0.87/10.0.0.87:9042
com.datastax.shaded.netty.channel.AbstractChannel$AnnotatedNoRouteToHostException: No route to host: 10.0.0.87/10.0.0.87:9042
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) ~[na:na]
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779) ~[na:na]
	at com.datastax.shaded.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:225) ~[scylla-driver-core-3.7.1-scylla-2-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:291) ~[scylla-driver-core-3.7.1-scylla-2-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:634) ~[scylla-driver-core-3.7.1-scylla-2-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:546) ~[scylla-driver-core-3.7.1-scylla-2-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:500) ~[scylla-driver-core-3.7.1-scylla-2-shaded.jar:na]
	at com.datastax.shaded.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:460) ~[scylla-driver-core-3.7.1-scylla-2-shaded.jar:na]
	at com.datastax.shaded.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:131) ~[scylla-driver-core-3.7.1-scylla-2-shaded.jar:na]
	at com.datastax.shaded.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[scylla-driver-core-3.7.1-scylla-2-shaded.jar:na]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: java.net.NoRouteToHostException: No route to host
	... 11 common frames omitted
com.datastax.driver.core.exceptions.WriteTimeoutException: Cassandra timeout during SIMPLE write query at consistency QUORUM (3 replica were required but only 2 acknowledged the write)
```

- Restore Monitor Stack command: `$ hydra investigate show-monitor 4133baf8-c193-41bc-a1c6-1f18c434701d`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=4133baf8-c193-41bc-a1c6-1f18c434701d)
- Show all stored logs command: `$ hydra investigate show-logs 4133baf8-c193-41bc-a1c6-1f18c434701d`

## Logs:
- **db-cluster-4133baf8.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/db-cluster-4133baf8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/db-cluster-4133baf8.tar.gz)
- **monitor-set-4133baf8.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/monitor-set-4133baf8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/monitor-set-4133baf8.tar.gz)
- **loader-set-4133baf8.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/loader-set-4133baf8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/loader-set-4133baf8.tar.gz)
- **sct-runner-4133baf8.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/sct-runner-4133baf8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/sct-runner-4133baf8.tar.gz)
- **parallel-timelines-report-4133baf8.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/parallel-timelines-report-4133baf8.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/4133baf8-c193-41bc-a1c6-1f18c434701d/20220429_130511/parallel-timelines-report-4133baf8.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-5.0/job/scale/job/scale-many-clients-4h-test/3/)

**Comment by raphaelsc** [comment]
10.0.1.85, which is the replacing node, triggered the problem

there are 50 keyspaces, with 1 table each

so 50 tables total

at peak, ~5600 sstables opened for each shard = 390k sstables across all 70 shards.

~5600 sstables / 50 keyspaces / shard = ~112 sstables / keyspace / shard


~400k sstables found in data dir:

```
402388 Index.db
402334 Data.db
402291 Digest.crc32
402240 Summary.db
402216 Filter.db
402216 CRC.db
402204 Statistics.db
402190 Scylla.db
402168 TOC.txt
704 TOC.txt.tmp
42 CompressionInfo.db
1 .lock
```

which should translate into ~805k file descriptors opened

OS metric reports 3.6 million inodes, which makes sense as ~805k sstables * 9 components for each = ~3.6 million

metric scylla_sstables_currently_open_for_writing shows # of sstables being currently written is always lower than 300

and scylla_sstables_total_open_for_writing shows that # of sstables even written doesn't go beyond ~400k

Luckily node_exporter has metric node_filefd_allocated, which shows that peak # of fd opened is ~928k.

ulimit shows the limit is: open files                      (-n) 1048576

**Comment by raphaelsc** [comment]
```
$ cat messages.log | grep "repair - Repair [[:digit:]]\+ out of [[:digit:]]\+ ranges" | grep "keyspace=" | awk '{print $19}' | sort -n | uniq -c
  52780 keyspace=keyspace12,
  52780 keyspace=keyspace15,
  52780 keyspace=keyspace17,
  52780 keyspace=keyspace20,
  52780 keyspace=keyspace26,
  52780 keyspace=keyspace28,
  52780 keyspace=keyspace4,
  48414 keyspace=keyspace40,
  52780 keyspace=keyspace47,

```

52780 / 70 = 754 ranges per shard

Sum of those ranges == 518654

With repair creating one sstable for each range, we have 518654 sstables, and 1037308 (data + index) file descriptors. enough to exhaust limit of 1048576.

For each keyspace with RF=3, we will create ~754 * SMP_COUNT.

This test has 50 keyspaces and 70 shards, so by completion, it would create 754×70×50=2639000 sstables. Which means  2639000 * 2 (data + index) = ~5.2 million descriptors. Which is a huge amount.

@asias I am thinking that we'll have to increase the granularity at which off-strategy is triggered by repair-based ops. So to allow off-strategy to reduce # of files. We could potentially track the # of sstables in a keyspace, and trigger off-strategy whenever it crosses some threshold.

**Comment by bhalevy** [comment]
> @asias I am thinking that we'll have to increase the granularity at which off-strategy is triggered by repair-based ops. So to allow off-strategy to reduce # of files. We could potentially track the # of sstables in a keyspace, and trigger off-strategy whenever it crosses some threshold.

I totally agree with that.  I already suggested in the past to trigger offstrategy compaction during repair if there are enough sstables already in the maintenance set.

Whether it be per table, per keyspace, per shard, or a global threshold, I'm not sure.  The file descriptor limit is global so it is fairly easy to come up with a sensible global or per-shard threshold (assuming the inter-shard variability would be low).
When the threshold is crossed we can then scan all tables and trigger offstrategy compaction in those tables that cross e.g. max_threshold, to filter out tables with too few sstables for which triggering compaction would be a waste.

**Comment by slivne** [comment]
@fgelcer  need to understand if this is a regression from 4.6 or not

@bhalevy if this is a regression we need to fix it prior to 5.0 going out.

**Comment by raphaelsc** [comment]
Commit 97bb2e47ff004b32b2d72f1b1f085710a14cb4e2, present in 4.6.0, enables repair-based replace by default. All tables use STCS, meaning no data segregation involved, and commit a4053dbb721707ddedb4910efc487042134ad013 is also available at 5.0 branch. So potentially not a regression but rather a preexisting issue. But I can be missing something, so let's wait for @fgelcer to confirm it.

**Comment by fgelcer** [comment]
> @fgelcer need to understand if this is a regression from 4.6 or not
> 
> @bhalevy if this is a regression we need to fix it prior to 5.0 going out.

@soyacz checked a 1TB job on 4.6 and saw the amount of open files seems normal, and he is now trying to reproduce it with the same job, running on 4.6

**Comment by soyacz** [comment]
I reexecuted the same test for 4.6 (although different set of nemesis got executed). During remove_node_then_add_node nemesis, problem exists but is few times smaller - we reached ~1M peak in number of open files, which was handled without issues.
![image](https://user-images.githubusercontent.com/18242288/166238607-14d26d0c-69ff-4d06-80bf-9c7f343d5b01.png)

For 4.6 1TB job mentioned by Fabio, there's a peak up to ~400k open files - which didn't cause any problems.

So it may be that problem got introduced in 4.6 and reinforced by some other change in 5.0.

**Comment by raphaelsc** [comment]
> I reexecuted the same test for 4.6 (although different set of nemesis got executed).

I need to understand exactly what changed between those two tests.

Please share with me the logs of the machine which # of files almost reached 1M

> During remove_node_then_add_node nemesis,

You're saying remove_node_then_add_node nemesis, which means bootstrap, and not replace, right? Repair-based op is only enabled for replace, so this 4.6 test is stressing something different. The issue says the problem was triggered with terminate_and_replace_node nemesis.

> problem exists but is few times smaller - we reached ~1M peak in number of open files, which was handled without issues. ![image](https://user-images.githubusercontent.com/18242288/166238607-14d26d0c-69ff-4d06-80bf-9c7f343d5b01.png)
> 
> For 4.6 1TB job mentioned by Fabio, there's a peak up to ~400k open files - which didn't cause any problems.
> 
> So it may be that problem got introduced in 4.6 and reinforced by some other change in 5.0.

**Comment by soyacz** [comment]
I confirm, on 4.6.0 issue is also present. In terminate_and_replace node we reach 4.6M files open.

**Comment by raphaelsc** [comment]
> I confirm, on 4.6.0 issue is also present. In terminate_and_replace node we reach 4.6M files open.

Good. Thanks @soyacz

**Comment by bhalevy** [comment]
The plan, as discussed with @raphaelsc, is to trigger off-strategy compaction when we finish with each sstable.
It will compact together the 754 sstables that are created and will compact them together into the main set in a single pass (using the incremental selector, since the sstables are disjoint).

@asias / @avikivity what do you think?

**Comment by raphaelsc** [comment]
> 

@bhalevy turns out the granularity of each range repair is keyspace, meaning we repair a range for all tables in a keyspace at once. That can be seen in the logs, where after repair is done with a range, 1 sstable is added to maintenance set for each of the tables belonging to the same keyspace.

So assuming a keyspace has N tables, where N is high, we'd probably have to trigger off-strategy earlier to put a cap on # of opened file descriptors. 

Assuming N = 100, S = 96, R = 754,  where S = smp count and R = # of ranges to be repaired

100 * 96 * 754 = 7238400 sstables

we'd have to pick R = 32 as off-strategy threshold, to reduce # of sstables significantly. 

for keyspace with less tables, we can be more relaxed about R as threshold, such that off-strategy can be more efficient

For example, with 1 table per keyspace, we could allow all 754 disjoint files to accumulate

With bootstrap / replace, it's easier to coordinate this as repair_info stores total # of ranges, and # of tables in the keyspace

We could today easily solve the problem triggered by the test, of many keyspaces, each with 1 or a few tables inside. Because as I said, the granularity is per keyspace, so repair could easily trigger off-strategy for all tables in keyspace once it's done with it, but the problem of keyspace with many tables would still remain.

Thoughts?

**Comment by bhalevy** [comment]
> > 
> 
> @bhalevy turns out the granularity of each range repair is keyspace, meaning we repair a range for all tables in a keyspace at once. That can be seen in the logs, where after repair is done with a range, 1 sstable is added to maintenance set for each of the tables belonging to the same keyspace.
> 
> So assuming a keyspace has N tables, where N is high, we'd probably have to trigger off-strategy earlier to put a cap on # of opened file descriptors.
> 
> Assuming N = 100, S = 96, R = 754, where S = smp count and R = # of ranges to be repaired
> 
> 100 * 96 * 754 = 7238400 sstables
> 
> we'd have to pick R = 32 as off-strategy threshold, to reduce # of sstables significantly.
> 
> for keyspace with less tables, we can be more relaxed about R as threshold, such that off-strategy can be more efficient
> 
> For example, with 1 table per keyspace, we could allow all 754 disjoint files to accumulate
> 
> With bootstrap / replace, it's easier to coordinate this as repair_info stores total # of ranges, and # of tables in the keyspace
> 
> We could today easily solve the problem triggered by the test, of many keyspaces, each with 1 or a few tables inside. Because as I said, the granularity is per keyspace, so repair could easily trigger off-strategy for all tables in keyspace once it's done with it, but the problem of keyspace with many tables would still remain.
> 
> Thoughts?

The only idea that come into my mind, is run the equivalent of regular compaction in the maintenance set (just with significantly higher min_threshold), until repair completes and then we run offstrategy compaction, since regular compaction that is confined to the maintenance set will reduce the number of sstables, but will keep them disjoint from each other.

**Comment by raphaelsc** [comment]
> > > 
> > 
> > 
> > @bhalevy turns out the granularity of each range repair is keyspace, meaning we repair a range for all tables in a keyspace at once. That can be seen in the logs, where after repair is done with a range, 1 sstable is added to maintenance set for each of the tables belonging to the same keyspace.
> > So assuming a keyspace has N tables, where N is high, we'd probably have to trigger off-strategy earlier to put a cap on # of opened file descriptors.
> > Assuming N = 100, S = 96, R = 754, where S = smp count and R = # of ranges to be repaired
> > 100 * 96 * 754 = 7238400 sstables
> > we'd have to pick R = 32 as off-strategy threshold, to reduce # of sstables significantly.
> > for keyspace with less tables, we can be more relaxed about R as threshold, such that off-strategy can be more efficient
> > For example, with 1 table per keyspace, we could allow all 754 disjoint files to accumulate
> > With bootstrap / replace, it's easier to coordinate this as repair_info stores total # of ranges, and # of tables in the keyspace
> > We could today easily solve the problem triggered by the test, of many keyspaces, each with 1 or a few tables inside. Because as I said, the granularity is per keyspace, so repair could easily trigger off-strategy for all tables in keyspace once it's done with it, but the problem of keyspace with many tables would still remain.
> > Thoughts?
> 
> The only idea that come into my mind, is run the equivalent of regular compaction in the maintenance set (just with significantly higher min_threshold), until repair completes and then we run offstrategy compaction, since regular compaction that is confined to the maintenance set will reduce the number of sstables, but will keep them disjoint from each other.

I think this could work. but we'd need to run the equivalent of major instead, as we want all input files to be placed into a single output file. To increase efficiency, output files of this intermediate compaction won't be picked by another "major", but only later by off-strategy, which will then perform reshape on still disjoint files and place them into main set. even if we pick a threshold of 32, we're adding only write amp of 1, as all data in maintenance set will only have to go through a single intermediate compaction. compaction is also essentially a copy operation, as its reader is able to incrementally select one file at a time. Sounds like we can go in this direction.

**Comment by asias** [comment]
Let's solve the issue in stages. 

First and easy thing we can do is triggering off strategy whenever a keyspace is finished. In RBNO, we process one keyspace at a time. 

Another thing we can do is to sync data using repair one table at a time and trigger off strategy when a table is done. 

Third thing we can do is to allow trigger off strategy automatically for a given table when accumulated sstables are more than some threshold.

**Comment by raphaelsc** [comment]
> Let's solve the issue in stages.
> 
> First and easy thing we can do is triggering off strategy whenever a keyspace is finished. In RBNO, we process one keyspace at a time.
> 
> Another thing we can do is to sync data using repair one table at a time and trigger off strategy when a table is done.

How hard is this change? Would we lose efficiency by serializing repair of tables in a keyspace? If we do this, we don't need the  threshold kicking off off-strategy earlier.

**Comment by asias** [comment]
> > Let's solve the issue in stages.
> > First and easy thing we can do is triggering off strategy whenever a keyspace is finished. In RBNO, we process one keyspace at a time.
> > Another thing we can do is to sync data using repair one table at a time and trigger off strategy when a table is done.
> 
> How hard is this change? Would we lose efficiency by serializing repair of tables in a keyspace? If we do this, we don't need the threshold kicking off off-strategy earlier.

The change is not trivial because inside repair we loop the ranges then the tables. 

We can still have parallelism if there are multiple ranges for the given table.

**Comment by bhalevy** [comment]
> > > Let's solve the issue in stages.
> > > First and easy thing we can do is triggering off strategy whenever a keyspace is finished. In RBNO, we process one keyspace at a time.
> > > Another thing we can do is to sync data using repair one table at a time and trigger off strategy when a table is done.
> > 
> > 
> > How hard is this change? Would we lose efficiency by serializing repair of tables in a keyspace? If we do this, we don't need the threshold kicking off off-strategy earlier.
> 
> The change is not trivial because inside repair we loop the ranges then the tables.
> 
> We can still have parallelism if there are multiple ranges for the given table.

@asias / @raphaelsc - so let's consider that only for the long term.
We still need to find a solution for 5.0.

Let's summarize the latest proposal we discussed:

- Introduce `_intermediate_sstables` in addition to `_main_sstables` and `_maintenance_sstables` in table.
- When the number of sstable in `_maintenance_sstables` crosses a threshold (e.g. max_threshold == 32), compact them together and place the output in `_intermediate_sstables`.  Note that the sstables in `_intermediate_sstables` are still disjoint from each other, and also with any sstable in `_maintenance_sstables`.
- When the number of sstable in `_intermediate_sstables` crosses the threshold, compact them together, keeping the output in `_intermediate_sstables`.
- When done, run offstrategy compaction on `_intermediate_sstables` + `_maintenance_sstables` like we do today, and place output sstables in `_main_sstables`.

@raphaelsc, @asias please review above proposal.
@avikivity are you ok with the above?

**Comment by asias** [comment]
I chatted with @raphaelsc , we both think adding another intermediate set is adding too much complexity.   Raphael proposed to compact directly inside the  _maintenance_sstables set if the number of files exceeds a threshold. 

In the mean while, I am working on changing repair internally to allow repair table by table so we can trigger off strategy and compact early.

**Comment by bhalevy** [comment]
> I chatted with @raphaelsc , we both think adding another intermediate set is adding too much complexity. Raphael proposed to compact directly inside the _maintenance_sstables set if the number of files exceeds a threshold.

This is possible, but we may want to use STCS buckets algorithm there to reduce write amplification.

> 
> In the mean while, I am working on changing repair internally to allow repair table by table so we can trigger off strategy and compact early.

great

**Comment by avikivity** [comment]
> The plan, as discussed with @raphaelsc, is to trigger off-strategy compaction when we finish with each sstable. It will compact 

Did you mean 'each table'?

> together the 754 sstables that are created and will compact them together into the main set in a single pass (using the incremental selector, since the sstables are disjoint).
> 
> @asias / @avikivity what do you think?

**Comment by bhalevy** [comment]
> > The plan, as discussed with @raphaelsc, is to trigger off-strategy compaction when we finish with each sstable. It will compact
> 
> Did you mean 'each table'?
> 

Yes, fixed.  thanks for spotting

**Comment by avikivity** [comment]
Ok, so it looks reasonable.

There's still the option that repair creates sstables faster then offstrategy cleans them. We can make non-system sstables_manager wait on sstable creation if the number of active sstables is too high.

**Comment by asias** [comment]
I have sent a PR: https://github.com/scylladb/scylla/pull/10551. With this patch, the first and second item [from](https://github.com/scylladb/scylla/issues/10462#issuecomment-1118058973
) are  implemented. It is going to help a lot even without the third item. Hopefully, the scale test is going to pas with the PR already.

**Comment by roydahan** [comment]
@slivne / @bhalevy shouldn't we mark this bug as 'high'?

**Comment by slivne** [comment]
@avikivity backport ping

**Comment by slivne** [comment]
not compared to 4.6 but it is a regression compared to 4.5

we enabled repair_base_node_operations for replace and that is whats
causing replace to fail in this case

**Comment by avikivity** [comment]
Ok. But the change is not trivial. Let's get some cycles on it.

**Comment by slivne** [comment]
we need to run a test on master with this patch and repair trying to see that this doesn't introduce an issue

**Comment by bhalevy** [comment]
> I have sent a PR: #10551. With this patch, the first and second item [from](https://github.com/scylladb/scylla/issues/10462#issuecomment-1118058973) are implemented. It is going to help a lot even without the third item. Hopefully, the scale test is going to pas with the PR already.

I think we can close this issue with e65b3ed50a1b4953eb57a5ed38438c06acd222e1, modulo testing of course.

> Third thing we can do is to allow trigger off strategy automatically for a given table when accumulated sstables are more than some threshold.

This can be considered separately as an optimization.
Not sure it will actually be needed.

**Comment by slivne** [comment]
@roydahan / @soyacz ping - did we run this on master after the fix from asias went in ?

**Comment by soyacz** [comment]
I just executed this nemesis on master and it looks open files count is under control:
![image](https://user-images.githubusercontent.com/18242288/171013041-8a74768b-e98c-4572-be05-28cafb69fe59.png)

**Comment by asias** [comment]
Which commit were you using? I want to ensure the fix is present and
effective.

**Comment by soyacz** [comment]
5.1.dev-0.20220525.dc5c9321fee9 with build-id 5b7f6260568fb8c82a0a426141fc01c4d21edf8a

**Comment by asias** [comment]
> 5.1.dev-0.20220525.dc5c9321fee9 with build-id 5b7f6260568fb8c82a0a426141fc01c4d21edf8a

I can confirm the version 5.1.dev-0.20220525.dc5c9321fee9 from the test contains the patches

```
commit 7a38b806bee6223dcb2564922f8e4810ef1d3406
Author: Asias He <asias@scylladb.com>
Date:   Thu May 12 10:11:31 2022 +0800

    repair: Trigger off strategy compaction after all ranges of a table is repaired


commit 3dc9a81d02a26d021ad1954d57b574ad4899561d
Author: Asias He <asias@scylladb.com>
Date:   Tue May 10 10:49:34 2022 +0800

    repair: Repair table by table internally
```

This suggest the patches to trigger off strategy earlier are effective.

@bhalevy It's up to you to prioritize the feature to compact inside maintenance set if number of sstables are getting big.

Can we close the issue for now?

**Comment by bhalevy** [comment]
> > 5.1.dev-0.20220525.dc5c9321fee9 with build-id 5b7f6260568fb8c82a0a426141fc01c4d21edf8a
> 
> I can confirm the version 5.1.dev-0.20220525.dc5c9321fee9 from the test contains the patches
> 
> ```
> commit 7a38b806bee6223dcb2564922f8e4810ef1d3406
> Author: Asias He <asias@scylladb.com>
> Date:   Thu May 12 10:11:31 2022 +0800
> 
>     repair: Trigger off strategy compaction after all ranges of a table is repaired
> 
> 
> commit 3dc9a81d02a26d021ad1954d57b574ad4899561d
> Author: Asias He <asias@scylladb.com>
> Date:   Tue May 10 10:49:34 2022 +0800
> 
>     repair: Repair table by table internally
> ```
> 
> This suggest the patches to trigger off strategy earlier are effective.
> 
> @bhalevy It's up to you to prioritize the feature to compact inside maintenance set if number of sstables are getting big.

Let's put this on hold until proven needed

> 
> Can we close the issue for now?

Yes, your solution seems good enough.

**Comment by slivne** [comment]
@avikivity we have a confirmation that asias patches fix the issue and they ave been tested can we do the backport

**Comment by avikivity** [comment]
Fixed by e65b3ed50a1b4953eb57a5ed38438c06acd222e1
